File: 0046-permutations/0046-permutations.java
class Solution {
    private final List<List<Integer>> perms = new ArrayList<>();
    private final List<Integer> curr = new ArrayList<>();
    private int[] elements;
    private boolean[] visited;

    public List<List<Integer>> permute(int[] nums) {
        elements = nums;
        visited = new boolean[nums.length];
        dfs(0);
        return perms;
    }

    private void dfs(int index) {

        if (index == elements.length) {
            perms.add(new ArrayList<>(curr));
            return;
        }

        for (int j = 0; j < elements.length; j++) {
            if (!visited[j]) {
                visited[j] = true;
                curr.add(elements[j]);
                dfs(index + 1);

                curr.remove(curr.size() - 1);
                visited[j] = false;
            }
        }
    }
}

File: 0046-permutations/0046-permutations.py
class Solution:
    def __init__(self):
        self.perms: List[List[int]] = []
        self.curr: List[int] = []
        self.elements: List[int] = []
        self.visited: List[bool] = []

    def permute(self, nums: List[int]) -> List[List[int]]:
        self.elements = nums
        self.visited = [False] * len(nums)
        self._dfs(0)
        return self.perms

    def _dfs(self, index: int) -> None:
        if index == len(self.elements):
            self.perms.append(self.curr[:])  # Copy current path
            return

        for j in range(len(self.elements)):
            if not self.visited[j]:
                self.visited[j] = True
                self.curr.append(self.elements[j])
                self._dfs(index + 1)
                self.curr.pop()
                self.visited[j] = False
